### PR TITLE
Bug 1170235 compat tables

### PR DESCRIPTION
**Pull request by stephaniehobson**
How to test:
- download this branch and compile your style sheets and compress assets
- create a macro called "Template:EmbedCompatTable" from this:  https://gist.github.com/darkwing/e6b94e895c7af6850682
- add the following function to "Template:mdn:common":  https://gist.github.com/darkwing/f1f004c098037bbe81fd
- create a document that uses the template like `{{ EmbedCompatTable('web-css-display') }}`

Still to do:
- [ ] Icon font not appearing (may be unrelated issue)
- [x] Space in the table slides open but the history drawer itself just appears.
- [x] Space in the table slides closed but the history drawer itself just dissipears.
- [x] The heading is also 'sliding open' when the history draw is opened in tablet/flex view. (I'll look at a fix)
- [x] Need explicit way to trigger thick border between support cells when medium switches.
- [x] Add `.bc-medium-border` to the macro
- [x] Add bug number to that one commit 
- [x] solve animation problem
- [x] decide if need waffle

Where should we identify problems with the macro? The contents, legend, and title of many of the icons/abbreviations need to be changed.

Supersedes #3258

### CODE REVIEW

File: media/stylus/components/compat-tables/bc-table.styl [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    /*
+    Flexbox for those lucky few who can handle it
+    - we only needs this to make equal height columns in the tablet display view
+    - ie is not among those that can handle this flexbox code,
+      but it doesn't supoort @support so we need to include that in the MQ
+    - all browsers which support @support also use unprefixed syntax
+    ----------------------------------------------------------------------  */
+
+    .bc-table,
+    .bc-table thead,
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-table.styl]: Flexible Box Layout Module not supported by: IE (10), IE Mobile (10)

File: media/stylus/components/compat-tables/bc-table.styl [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-table.styl]: Flexible Box Layout Module not supported by: IE (10), IE Mobile (10)

File: media/stylus/components/compat-tables/bc-table.styl [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+bc-table-tablet-display-flex() {
+    /*
+    Flexbox for those lucky few who can handle it
+    - we only needs this to make equal height columns in the tablet display view
+    - ie is not among those that can handle this flexbox code,
+      but it doesn't supoort @support so we need to include that in the MQ
+    - all browsers which support @support also use unprefixed syntax
+    ----------------------------------------------------------------------  */
+
+    .bc-table,
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-table.styl]: Flexible Box Layout Module not supported by: IE (10), IE Mobile (10)

File: media/stylus/components/compat-tables/bc-legend.styl [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/*
+legend
+====================================================================== */
+
+.bc-legend {
+    -moz-column-count: 2;
+    -webkit-column-count: 2;
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-legend.styl]: CSS3 Multiple column layout not supported by: Firefox (37,38), Chrome (42,43), Safari (8,7.1), Opera (29,30), iOS Safari (8,8.1-8.3)

File: media/stylus/components/compat-tables/bc-table.styl (hunk 3 above)
Comment (stephaniehobson) [inline media/stylus/components/compat-tables/bc-table.styl]: Discord is pretty cool! 

@openjck These comments are appearing a few lines higher than they probably should. Considering it's source mapped that's still pretty cool.

**Comment by darkwing** [comment]
OK, so assuming the font thing is a local issue, the one thing I don't like with all of this is the `setTimeout` we're using to do the table / hide effect.

@stephaniehobson Can we use CSS animation delay to avoid the `setTimeout`?  If not, we may be better off going another route, one maybe more JS-based.

Let me know what you think.

File: media/stylus/components/compat-tables/bc-legend.styl [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/*
+legend
+====================================================================== */
+
+.bc-legend {
+    vendorize(column-count, 2);
+    page-break-inside: avoid;
+    webkit-column-break-inside: avoid;
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-legend.styl]: CSS3 Multiple column layout not supported by: Firefox (37,38), Chrome (42,43), Safari (8,7.1), Opera (29,30), iOS Safari (8,8.1-8.3)

File: media/stylus/components/compat-tables/bc-table.styl (hunk 2 above)
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-table.styl]: Flexible Box Layout Module not supported by: IE (10), IE Mobile (10)

File: media/stylus/components/compat-tables/bc-table.styl (hunk 3 above)
Comment (groovecoder) [inline media/stylus/components/compat-tables/bc-table.styl]: FWIW: https://github.com/mdn/discord/issues/61

**Comment by groovecoder** [comment]
I don't see any waffle in here, but is this code innocuous enough to merge & push to production? Or should we waffle (part) of it?

**Comment by stephaniehobson** [comment]
I think it's safe to merge unwaffled, the files don't do anything unless we include them (exept for a few small organization changes I made to the stylus files that shouldn't affect anything). @darkwing do you agree?

**Comment by stephaniehobson** [comment]
@darkwing we can't use transitionend to get around the CSS delay because the browser support isn't good enough. I think the best bet is to use jQuery to handle the entire animation.

File: media/stylus/components/compat-tables/bc-table.styl (hunk 2 above)
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-table.styl]: Flexible Box Layout Module not supported by: IE (10), IE Mobile (10)

File: media/stylus/components/compat-tables/bc-legend.styl [hunk 6]
```diff
@@ -0,0 +1,26 @@
+/*
+legend
+====================================================================== */
+
+.bc-legend {
+    vendorize(column-count, 2);
+    page-break-inside: avoid;
```
Comment (mdn-services) [inline media/stylus/components/compat-tables/bc-legend.styl]: CSS3 Multiple column layout not supported by: Firefox (37,38), Chrome (42,43), Safari (8,7.1), Opera (29,30), iOS Safari (8,8.1-8.3)

**Comment by darkwing** [comment]
We aren't currently waffled because:

https://github.com/mozilla/kuma/pull/3344/files#diff-c9430f01b685e1d90aed5ca7b9babb31R297

If no tables are found, we don't load anything.

I don't think we can do this by traditional waffle.  I think we (unfortunately) may need a "router" kumascript macro in which we decide (manually) whether to load the old compat table code or the new compat table code.  I say this because the HTML used to build the tables now is much, much different than what it will be with this update.

What sucks about using kumascript to decide which tables to use is that the kumascript result gets cached, and thus each doc will need a hard-refresh to re-render.

**Comment by darkwing** [comment]
Before we merge this PR, we should really have a meeting where we discuss where we are on the dev side (both python and front-end), and how we plan to roll this out.  If we don't plan on rolling this out for a while, I would *not* merge this PR; we should learn from the past (i.e. contrib bar) and wait until we're willing to give it max attention and somewhat immediate implementation.  I don't anticipate much bitrot with this since it's mostly all new files.

**Comment by jwhitlock** [comment]
We should have a meeting - I've typed and retyped my response a few times.

This is how I think it will go down:
1) Testing in developer boxes, using copies of MDN pages against browsercompat.herokuapp.com. Go through feedback until dev are mostly happy with the implementation,
2) Merged this branch into master, ship assets to stage and production.  Kumascript is unused on production (is this enforceable?).  Select pages are manually enabled on staging by replacing content with new kumascript.  Staff (writers, QA, management, etc.) view and comment, we play with refresh scenarios, etc.
3) Ship a production-ready version of API server.  Ship contribution workflow site.  Load test, QA test, etc.
4) Manually migrate some pages on production, load test, QA test, etc.
5) Actively convert pages on production.

I'm focusing on some other tasks, but when I'm done, I'll take a look at this branch on my dev box.

**Comment by stephaniehobson** [comment]
To begin with we can just not create the template on production, that will keep it unused ;) We will also lose the template when database updates are done to staging.

We will probably need some kind of change management while we are developing the kumascript macro, not sure where that should live.

**Comment by groovecoder** [comment]
Tasks for merging this:
* [ ] @darkwing - double-check that we won't add extra .css. and .js calls on production

Tasks for starting showing tables on the stage server:

* [ ] @darkwing - create KS template that accepts a feature slug and returns the BrowserCompat API response payload
  * error-handling?
  * link to contribution workflow?
  * how much JS in kumascript template vs. client-side JS?
* [ ] @jwhitlock - pick some example pages (with @stephaniehobson) to convert on stage to use new KS template

Tasks to start now to anticipate production deployment:
* [ ] @hoosteeno - add "BrowserCompat" component to our bugzilla product
* [ ] @groovecoder file bugs for ...
* [ ] Investigate how - and how often - updated browsercompat data will get to MDN page display via KumaScript
  * BrowserCompat "push" update ping (webhook?) to MDN to re-render pages?
* [ ] Consider effects of resetting/updating each related (e.g., languages') page's table display
  * separate celery queue just for these

**Comment by groovecoder** [comment]
Updated comment to be just the 1 task blocking this PR from merge.

**Comment by groovecoder** [comment]
Note: Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1187341 too

**Comment by darkwing** [comment]
@stephaniehobson Can you update the commit message with a bug number?

**Comment by darkwing** [comment]
I've updated the instructions at the top of this PR with macro code to make everything work! :)

**Comment by stephaniehobson** [comment]
So @groovecoder and I both tried to add the bug number and couldn't so we decided it's okay not to have a bug number, just this once.

**Comment by groovecoder** [comment]
:+1:
giphy

**Comment by groovecoder** [comment]
I tried to get into this quickly on a db dump, but I think that messed it up. So, I'm going to check out https://github.com/mozilla/kuma/pull/3352 first.

**Comment by groovecoder** [comment]
Alright, I've finally got something going. But it looks like a bunch of assets might be missing that's breaking the display?
embed-compat-table-local
Seen this before? Any easy fix?

**Comment by darkwing** [comment]
@groovecoder I've updated the KumaScript template (https://gist.github.com/darkwing/e6b94e895c7af6850682) to the newest image URL.  Thanks!

**Comment by groovecoder** [comment]
It's a little better, but I'm still not seeing images. And shouldn't there be some collapsible behavior going on?

**Comment by darkwing** [comment]
The CSS and JS resources are lazyloaded so you need to compress_assets.  Also, if you look at the path set in the kumascript template and the location of the image file in the "FILES" view of this PR, you'll see they match. Are you sure you have the updated branch checked out?

**Comment by groovecoder** [comment]
Ah, yeah I checked out the branch again, ran `compress_assets` and things look much better. It's still failing to load the compat icons font though?

**Comment by darkwing** [comment]
The font issue is something I noticed as well -- I was hoping we could put this on staging to see if it's a local kuma issue, since the font works fine everywhere outside our local VMs

**Comment by groovecoder** [comment]
Makes sense. Happy to merge now. w00t!

**Pull request merged**